tgui/packages/tgui/components/DraggableControl.js

```diff
@@ -0,0 +1,280 @@
+      const {
+        dragging,
+        value,
+        internalValue,
+      } = this.state;
+      document.body.style['pointer-events'] = 'auto';
+      clearTimeout(this.timer);
+      clearInterval(this.dragInterval);
+      this.setState({
+        dragging: false,
+        editing: !dragging,
+        origin: null,
+      });
+      document.removeEventListener('mousemove', this.handleDragMove);
+      document.removeEventListener('mouseup', this.handleDragEnd);
+      if (dragging) {
+        this.suppressFlicker();
+        if (onChange) {
+          onChange(e, value);
+        }
+        if (onDrag) {
+          onDrag(e, value);
+        }
+      }
+      else if (this.inputRef) {
+        const input = this.inputRef.current;
+        input.value = internalValue;
+        // IE8: Dies when trying to focus a hidden element
+        // (Error: Object does not support this action)
+        try {
+          input.focus();
+          input.select();
+        }
+        catch {}
+      }
+    };
+  }
+
+  render() {
+    const {
+      dragging,
+      editing,
+      value: intermediateValue,
+      suppressingFlicker,
+    } = this.state;
+    const {
+      animated,
+      value,
+      unit,
+      minValue,
+      maxValue,
+      format,
+      onChange,
+      onDrag,
+      children,
+      // Input props
+      height,
+      lineHeight,
+      fontSize,
+    } = this.props;
+    let displayValue = value;
+    if (dragging || suppressingFlicker) {
+      displayValue = intermediateValue;
+    }
+    // Setup a display element
+    // Shows a formatted number based on what we are currently doing
+    // with the draggable surface.
+    const renderDisplayElement = value => (
+      value + (unit ? ' ' + unit : '')
+    );
+    const displayElement = (
+      animated && !dragging && !suppressingFlicker && (
+        <AnimatedNumber
+          value={displayValue}
+          format={format}>
+          {renderDisplayElement}
+        </AnimatedNumber>
+      ) || (
+        renderDisplayElement(format
+          ? format(displayValue)
+          : displayValue)
+      )
+    );
+    // Setup an input element
+    // Handles direct input via the keyboard
+    const inputElement = (
+      <input
+        ref={this.inputRef}
+        className="NumberInput__input"
+        style={{
+          display: !editing ? 'none' : undefined,
+          height: height,
+          'line-height': lineHeight,
+          'font-size': fontSize,
+        }}
+        onBlur={e => {
+          if (!editing) {
+            return;
+          }
+          const value = clamp(e.target.value, minValue, maxValue);
+          this.setState({
+            editing: false,
+            value,
+          });
+          this.suppressFlicker();
+          if (onChange) {
+            onChange(e, value);
+          }
+          if (onDrag) {
+            onDrag(e, value);
+          }
+        }}
+        onKeyDown={e => {
+          if (e.keyCode === 13) {
```

Comment:
Any use of `keyCode` seriously, *seriously* needs either reference to a constant from a constants file or, at the minimum, an inline comment.

Reply:
Agreeable.